Unidirectional topic operator (#76)
strimzi/proposals: 050-unidirectional-topic-operator.md

---

Similarly to how it is possible to "manage" an existing unmanaged topic in Kafka, we will also provide a mechanism for "unmanaging" a managed topic, via a `spec.managed` flag.
This will allow a `KafkaTopic` to be deleted **without** the topic in Kafka being deleted.
This can be useful operationally, for example to change the `metadata.name` of a `KafkaTopic`.

> An alternative to `spec.managed` would be "explicit deletion" via a `spec.delete` flag.
> This alternative was rejected because it's incompatible with gitops.

---

#### Example 4: `KafkaTopic` deletion WITHOUT propagation to Kafka

> 1. Assume the topic is currently Kube-managed
> 
>     ```yaml=
>     metadata:
>       generation: 123
>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>     spec:
>       # ...
>     status:
>       observedGeneration: 123 # == generation => it is Kube-managed
>     ```
> 
> 2. The user changes the new `spec.managed` field to `False` so that the topic is no longer Kube-managed.
> 
>     ```yaml=
>     metadata:
>       generation: 124
>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>     spec:

> >       managed: False # do not synchronize to Kafka

Author: The user operator uses a `strimzi.io/pause-reconciliation` annotation to do a similar thing. We should use that (if the semantics are the same), or be explicit about why we've not using it (if the semantics of `spec.managed` are necessarily different)

Reviewer: This annotation is used in Cluster Operator as well. When set, the condition in `.status` is set to `Paused`.

Author: What happens if you delete a paused CR?

Reviewer: I actually don't know for sure about User Operator. I expect the deletion will proceed since eventually in the next timer based reconciliation, it will not remember anymore that this was paused user. For the resources such as `Kafka`, the Kube GC will wipe the resources right away.

Author: Since the UO doesn't use finalizers if a paused `KafkaUser` is deleted then the user will be removed from Kafka. So the behaviour seems pretty similar. Maybe we should use the annotation instead to be consistent. That would also fit nicely with using the Paused annotation instead of removing the `spec.topicName`. WDYT?

Reviewer: You mean instead of having the `.spec.managed` field to have some annotation `strimzi.io/managed: false`? That works for me, but not sure it makes a big difference. 

Or do you want to use the paused-reconciliation annotation for this but not delete the topic from Kafka? TBH, it would work for me as well. But it would be a different behavior from the User or Cluster operator as there will the user or cluster be really deleted.

The TO obviously works differently as it will not have any _reconcile all topics_ method as it doe snot manage all topics. In the UO, it is not the deletion of the paused KafkaUser which triggers the deletion. It is the periodical reconciliation and having the CRs as a single source of truth. Similarly, It is not the CO deleting the Kafka cluster which is paused - it is Kubernetes GC. So code-wise, the TO behavior would be the same. But from the user perspective it would differ.

Author: On balance: 

* I think the difference between paused and unmanaged is fundamental and important, so let's not use `strimzi.io/pause-reconciliation`.
* But it's probably more consistent to use an annotation for managed than to put it in the `spec`. So I'll change it to `strimzi.io/managed` with a boolean value, defaulting to true.
* I thought too about the `Paused` condition and whether to use that or a `Managed=false` condition. But the way the operator works it will simply apply the different logic on the next reconciliation: It's not a piece of state that needs to be reconciled. i.e. it's just not possible for a resource with the annotation to behave transiently in a way that's incompatible with the annotation.

>     status:
>       observedGeneration: 123 # != generation => it is Kube-managed
>     ```
> 
> 3. The user waits until the operator has observed that change, indicated by the `status.observedGeneration` matching the `metadata.generation`:
> 
>     ```yaml=
>     metadata:
>       generation: 124
>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>     spec:
>       managed: False # do not synchronize to Kafka

> >     status:

Author: We should also remove the `spec.topicName` to help indicate that there might not be a topic in Kafka for this `KafkaTopic`.

Reviewer: You cannot do that, because if you remove `.spec.topicName` you might not later know what the actual topic name was, or?

Author: That would be fine, actually. 

1. `spec.managed=false` 
2. then changing the `spec.topicName` 
3. then `spec.managed=true`

 just means you changed which topic in Kafka is being managed by the resource, just like it would if step 2 was deletion and recreation.

Reviewer: Yes. But the point is that:
* You might fight with some GitOps tooling which would set the `spec.topicName` back
* Assuming the GitOps tooling did not reverse your change, the user will be unable to just edit the spec and set `spec.managed=true` to recreate the topic / have it managed again.

The procedure you describe is fine if done by the user. But I do not think the UTO should edit the `.spec` this way.

The operator will emit a `WARN` level log on startup if autocreation is enabled. 

Alternatively: App could be written to wait for topic existency, or use an init container to do the same.

## Affected/not affected projects

* The existing bidirectional Topic Operator would be deprecated and removed once ZooKeeper-based clusters are no longer supported.
* A new unidirectional Topic Operator would be provided as a replacement, which users could migrate to before ZooKeeper-based clusters are no longer supported.
* The schema of `KafkaTopic` would change
* The schema of `Kafka` would also change, to support deployment of the new operator, but the details of that are not defined in thisproposal.
* The CRD changes would be relfected in the `api` module.

## Compatibility

// Call out any future or backwards compatibility considerations this proposal has accounted for.

To migrate, existing BTO users would have to:

1. Review how they're using the BTO and whether they require birdirectionality
2. Undeploy the BTO (they can retain their existing `KafkaTopics`)
3. Deploy the UTO

> ### Disadvantages

Reviewer: The use of finalizers should be listed in the disadvantages as well I think. It is something that might have its justification, but it also brings issues to deletions.

Author: Yes, done.

  * `status.topicName` is used to detect a change to the `spec.topicName`, which results in an error
  * `metadata.generation == status.observedGeneration` implies that the operator has reconciled the latest change to the `spec`.
    Whether that reconciliation was successful is indicated by the `Ready` condition having status `True`.

Some new fields will be introduced in the following text.

### The "single resource" principal

To avoid ambiguity we will continue to require that a single `KafkaTopic` is used to manage a single topic in a single Kafka cluster.

The fact that there are legal topic names (in Kafka) that are not legal resource names (in Kube) provided the motivation for supporting `spec.topicName`, so that a topic in Kafka could be managed by a `KafkaTopic` with a different `metadata.name`.
As such it's not possible for the Kube apiserver to enforce the single resource principal because multiple `KafkaTopic` resource might refer to the same topic (i.e. the uniqueness constraint on `metadata.name` is insufficient).
It is therefore left to the operator to detect if there are multiple `KafkaTopics` which point to the same topic in Kafka.

> To do this, the operator will keep an in-memory mapping of topic name to (namespace, name)-pairs.

Author: We could do that, but:

* Admission webhooks are essentially global to the Kube cluster, whereas there can be multiple independent TO's in a Kube cluster. To support multiple TOs the webhook would need a global view over all those TOs and how they mapped to Kafka clusters, because it's fine to have two `KafkaTopics` with `spec.topicName=foo` if they're in different Kafka cluster.
    - even if we felt this was worth the complexity, consider too that this needs to cope with TOs coming and going, being reconfigured. So the incremental complexity for the full dynamic case is significantly more than the base complexity of the purely static case.
* It's another process/pod/deployment:
    - it would _still_ need an in-memory map AFAICS, only this time potentially a much larger map, because it's now keyed on (clusterId, topicName)-pairs and covers all the Kafka clusters in the Kube cluster. Probably it's roughly the same memory usage in total, just it's now lumped in one process, but there doesn't seem to be an _advantage_ wrt memory usage overall. 
    - it creates a new failure mode where the TO is running but the WebHook is not configured. So it doesn't provide as good safety guarantee as making the same check in the TO itself.

> 
> ```yaml=
>     policy:
>     - namespace: team.a
>       topicNamePrefixes: 
>       - foo-app.
>       - bar-app-
>       topicNames:
>       - config-foo
>     - namespace: team.b
>       topicNamePrefixes
>       - quux-app.
>       topicNames:
>       - config-quux
>     - namespace: kafka.cluster.admins
>       otherTopics: True
> ```
> 
> Thus:
> 
> - Topics `foo-app.x` and `foo-app.y` can only be managed by `KafkaTopics `in namespace `team.a` (due to the `foo-app.` prefix), likewise `bar-app-x` and `bar-app-y` (due to the `bar-app-` prefix). Additionally `config-foo` can be only managed from this namespace (due to the `topicNames` list).
> - Topics `quux-app.x` and `quux-app.y` can only be managed by `KafkaTopics `in namespace `team.b` (due to the `quux-app.` prefix).
> Additionally `config-quux` can be only managed from this namespace (due to the `topicNames` list).
> - Topics `config-gee`, `__consumer_offsets`, and `__transaction_state` could only be managed from the `kafka.cluster.admins` namespace.
>
> Rules for legal policies:
> 
> * All `topicNamePrefixes` must be disjoint (i.e. not given prefix is a prefix of any other prefix).

> > * Not `topicNamePrefix` maybe be a prefix of of any listed `topicNames`.

Author: ```suggestion
> * No given `topicNamePrefix` maybe be a prefix of of any listed `topicNames`.
```

<tr><td>STRIMZI_KEYSTORE_LOCATION			<td>Unchanged
<tr><td>STRIMZI_KEYSTORE_PASSWORD			<td>Unchanged
<tr><td>STRIMZI_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM	<td>Unchanged
<tr><td>STRIMZI_SASL_ENABLED				<td>Unchanged
<tr><td>STRIMZI_SASL_MECHANISM				<td>Unchanged
<tr><td>STRIMZI_SASL_USERNAME				<td>Unchanged
<tr><td>STRIMZI_SASL_PASSWORD				<td>Unchanged
<tr><td>STRIMZI_SECURITY_PROTOCOL			<td>Unchanged
<tr><td>STRIMZI_ZOOKEEPER_CONNECT			<td>Dropped
<tr><td>STRIMZI_ZOOKEEPER_SESSION_TIMEOUT_MS		<td>Dropped
<tr><td>TC_ZK_CONNECTION_TIMEOUT_MS			<td>Dropped
<tr><td>STRIMZI_REASSIGN_THROTTLE			<td>Dropped, see note 2
<tr><td>STRIMZI_REASSIGN_VERIFY_INTERVAL_MS		<td>Dropped, see note 2
<tr><td>STRIMZI_TOPIC_METADATA_MAX_ATTEMPTS		<td>Dropped
<tr><td>STRIMZI_TOPICS_PATH				<td>Dropped
<tr><td>STRIMZI_STORE_TOPIC				<td>Dropped
<tr><td>STRIMZI_STORE_NAME				<td>Dropped
<tr><td>STRIMZI_APPLICATION_ID				<td>Dropped
<tr><td>STRIMZI_STALE_RESULT_TIMEOUT_MS			<td>Dropped
<tr><td>STRIMZI_USE_ZOOKEEPER_TOPIC_STORE		<td>Dropped
</table>

Notes:

1. `STRIMZI_NAMESPACE` is retained for backwards compatibility and configures the operator with a policy allowing all topics in the given namespace.

>    `STRIMZI_POLICY_JSON` configures a policy as JSON string.

Author: I don't have a strong preference. I went with JSON because I felt a suspicious of env vars containing newlines  which YAML would require (it seems likely that something, somewhere, eventually is going to do something with the significant whitespace).

<tr><td>gauge		<td>strimzi.resource.state		<th>Unchanged

### Emitted Kube `Events`

Let's consider all the cases in which the BTO emits `Event` resources to the Kube apiserver:

* For errors when accessing the topic store. 
  For UTO there is no topic store, so this cannot happen.
* For errors when creating partitions. 
  For UTO ApiExceptions from the Admin client will be reported via `Ready` condition.
  For other errors application logging and metrics are considered sufficient.
* For errors when changing the configuration of topics in Kafka
  For UTO ApiExceptions from the Admin client will be reported via `Ready` condition.
  For other errors application logging and metrics are considered sufficient.
* For errors when processing a ZooKeeper znode watch.
  UTO doesn't use znode watches, so this cannot happen.
* On attempts to change a `KafkaTopic`'s `spec.topicName`.
  For UTO this will be reported via the status only.
* On detection of conflicting changes between Kube and Kafka
  For UTO such conflicts cannot happen.
* On attempts to decreate the number of partitions.
  For UTO this will be reported via the status only.

> Thus, the UTO will emit no events.

Reviewer 2: I find that Kubernetes events provide a different dimension to native kube applications observability than logging, metrics, and status conditions. 
1. **Compared to logs & Metrics** : UTO logs and metrics are meant to be infrastructure status targeting the cluster administrators. In contrast, namespaced Kube events target the CR authors who might not have access to the CO logs and metrics. This becomes increasingly important for operators that provide some form of namespace policies or multi-tenancy-like operation mode such as the UTO. 
2. **Compared to status** Kube events provide more of a historical view of state change or error progression. For E.g. a current `Ready` resource with 0 events is different from a `Ready` resource with associated Kube events that show e.g. conflict -> ready -> conflict. Especially for a case when another resource with a conflicting name is created that ends up in **both** resources becoming not ready.

Author: @devguyio that's true, but I don't see any of those existing BTO events really requiring an `Event`. The one example you give (`ResourceConflict`) can only happen as a result of someone creating a conflicting `KafkaTopic` in the same namespace as an existing one. Presumably they're looking for the `Ready` in that case and will resolve the problem more or less immediately. Those resources being in conflict doesn't have any impact on how the topic is being used by Kafka applications. So I just don't seem much value in the historical state change in that case.

I think there might be a role for `Events` in the future. For example if/when we add support for RF change then when data movement is required it can take a long time. There seems a stronger case for using events to notify about those state changes (perhaps in addition to a 2nd condition with `type: ReplicasReady` perhaps), because we don't expect anyone to waiting on the outcome. But I don't want to say that in _this_ proposal, because I don't want to bite off RF change at the same time as a whole new operator.

Reviewer 2: Yeah this makes sense to me.

    Whether that reconciliation was successful is indicated by the `Ready` condition having status `True`.

Some new fields will be introduced in the following text.

### The "single resource" principal

To avoid ambiguity we will continue to require that a single `KafkaTopic` is used to manage a single topic in a single Kafka cluster.

The fact that there are legal topic names (in Kafka) that are not legal resource names (in Kube) provided the motivation for supporting `spec.topicName`, so that a topic in Kafka could be managed by a `KafkaTopic` with a different `metadata.name`.
As such it's not possible for the Kube apiserver to enforce the single resource principal because multiple `KafkaTopic` resource might refer to the same topic (i.e. the uniqueness constraint on `metadata.name` is insufficient).
It is therefore left to the operator to detect if there are multiple `KafkaTopics` which point to the same topic in Kafka.

To do this, the operator will keep an in-memory mapping of topic name to (namespace, name)-pairs.
This will allow us to detect the case where multiple resources are attempting to manage the same topic in Kafka.

> When this happens both resources' `Ready` status will change to `False` with suitable `reason`.

Author: I guess we could do that. It's not impossible for the creationTimestamps to be the same, but very unlikely. I think this will work, but I'll confirm.

#### Changing `spec.relicas`

This will NOT be supported using as part of this proposal.

#### Recreating topics in the Kafka cluster

This WILL be supported.

#### Deployment of application with `KafkaTopics`

Consider an application deployed via a manifest containing a `Deployment` and some `KafkaTopics`.
If the Kafka cluster is configured with `auto.create.topics.enable` there is a race condition between:

1. the operator reacting to the new `KafkaTopics`.
2. the application starting and triggering autocreation of the topics.

**If the operator wins:** then auto-creation will not happen, and the user's intent is honoured.

**If the application wins:** then the topics will be created using the default configuration for the cluster. The operator will later reconcile the `KafkaTopics` which will result in reconfiguration the default configuration for auto-created topics differs from the `KafkaTopic.spec`. That reconfiguration will either succeed, or fail (e.g. because a `KafkaTopic` has a lower `spec.partitions` than the auto-created topic).

To avoid this it is **recommended** that the Topic Operator not be used with cluster where `auto.create.topics.enable` is `true`. 

> The operator will emit a `WARN` level log on startup if autocreation is enabled. 

Reviewer 3: I've seen this happening with users more than one time and recovering from bad cluster defaults requires service disruption. In addition to the log on startup, can we also emit a Kube event with this warning?

Author: Do we know that users are more likely to notice a Kube `Event` than a warning in the logs?  

Emitting a Kube `Event` doesn't avoid any service disruption caused by changing the default cluster behaviour. It only makes the likely mistake more noticeable. And it's not _necessarily_ a mistake: It's only a problem if the user isn't creating their topics prior to application deployment. And even if it _is_ a mistake, it's not necessarily consequential (we expect the topic would get reconfigured almost immediately, so it's only a problem for certain changes, like RF change, or changing the #partitions). 

I'm not opposed to doing this, but I'm not completely convinced it would actually make a difference.

Reviewer 3: > It's only a problem if the user isn't creating their topics prior to application deployment. 

Agree, but this is exactly the situation depicted in the gitops vs application race, which is fairly common nowadays. I think we could be fine with the log warning and a documentation paragraph/section describing potential issues with topic auto autocreation, maybe in relation to few common use cases.

# A unidirectional topic operator

Replace the existing bidirectional topic operator with a unidirectional operator.

## Current situation

The Strimzi Topic Operator provides a Kubernetes API for viewing and modifying topics within a Kafka cluster.
It deviates from the standard operator pattern by being bidirectional.
That is, it will reconcile changes to a `KafkaTopic` resource both to _and from_ a Kafka cluster.
The bidirectionality means applications and users can continue to use Kafka-native APIs (such as the `Admin` client) and tooling (e.g. the scripts provided by Apache Kafka) as required.
The KafkaTopic's `spec` will be updated by the operator to reflect those changes.

The Strimzi community often abbreviates the topic operator as TO, but to avoid ambiguity in this document we will refer to the existing bidirectional topic operator as the BTO, and the proposed unidirectional topic operator as the UTO.

The BTO makes use of ZooKeeper znode watches to know about changes to topic state within the Kafka cluster.

> But Apache Kafka is removing its ZooKeeper-dependence which means that the BTO won't work with Kafka clusters that use it's replacement, known as KRaft.

Reviewer 4: ```suggestion
But Apache Kafka is removing its ZooKeeper-dependence which means that the BTO won't work with Kafka clusters that use its replacement, known as KRaft.
```

* It doesn't *enforce* a single source of truth, allowing for changes done directly in Kafka to conflict with declarative deployments. 

Moreover, the BTO doesn't do a good job with item 2. either:

* It only surfaces a subset of topic state through the Kube API (it doesn't include partition leadership/ISR for example)
This is intentional (it doesn't make sense to overload etcd to present this kind of dynamic information), but it undermines the suitability of the BTO for this use case.

For these reasons we are proposing the UTO.

## Proposal

### Goals

* A unidirectional topic operator which is KRaft compatible

### Non-goals

* Supporting replication factor change (this could be added later)

### Maintaining compatibility of the `KafkaTopic` custom resource

We would remain schema-compatible with the existing `KafkaTopic` resource at the API level. 
This means that all the existing `spec` sections could be handled by the new operator and all existing fields of the `status` section would still be present for applications that consume the status.

> However, the semantics of a `KafkaTopic` would change incompatibly: The `spec` would no longer be updated when Kafka-side configuration changed.

Reviewer: When you say that the semantics of the `KafkaTopic` changes, I would imagine that the fields remain the same but have a different meaning. I do not think that is the case here -> the semantics of the KafkaTopic custom resource is still the same, and all the fields still have the same meaning. The backward incompatible change is from my point of view in the TO itself, not in the CR. So I think this could be a bit confusing.

>     ```yaml
>     kind: KafkaTopic
>     metadata:
>       generation: 1
>       name: foo
>     status:
>       topicName: foo
>       observedGeneration: 1
>       conditions:
>       - type: Ready
>         status: False
>         reason: ResourceConflict
>         message: Managed by multiple KafkaTopic resources: some-namespace/bar
>         lastTransitionTime: 20230301T103000Z
>     ```
>     The reconciliation ends
>
> 3. Otherwise the operator attempts to it create the topic in Kafka. 
>
> 4. If the `CreateTopics` request fails due to `TOPIC_ALREADY_EXISTS` then the operator updates the topic in Kafka to reflect the `spec` using `incrementalAlterConfigs()`, and/or `createPartitions()`
>
> 5.The `status` is updated to reflect the Admin client requests. In the path case it would look like this:
> ```yaml=
> metadata:
>   generation: 1
>   name: foo
> spec:
>   topicName: foo
>   partitions: 12
>   replicas: 50
> status: 
>   observedGeneration: 1
>   topicName: foo
>   conditions:
>   - type: Ready
>     status: True
>     lastTransitionTime: 20230301T103000Z
> ```
> 
---

> 

Reviewer: So many empty lines are probably not needed.

        reason: NotSupported
        message: Changing spec.replicas is not supported by the operator
        lastTransitionTime: 20230301T103000Z
    ```

> Note that both these `NotSupported` reasons can arise both from changes being made to the `KafkaTopic`, or changes made directly in Kafka.
> For example, the UTO cannot tell whether the `spec.partitions` was decreased, or whether it was unchanged and the conflict arises because someone increased the number of partitions directly in Kafka so that it looks like the `KafkaTopic` is requesting a decrease. 
> In any case, when the user takes action to fix the problem the `KafkaTopic` will (eventually) get reconciled and the `status` will revert to `Ready`.

### Topic deletion

By default we will use a [Kube finalizer](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/) for deletion.
This means that deletion via the kube REST API will first mark the resource as scheduled for deletion by setting the `metadata.delectionTimestamp`, allowing the operator to handle the deletion and then update the `KafkaTopic`, removing its `metadata.finalizer` so that the resource is actually removed from `etcd`.

> Without a finalizer then the following would be posible:

Reviewer 4: ```suggestion
Without a finalizer then the following would be possible:
```

The operator cannot use the existence of a topic in Kafka to infer that the `KafkaTopic` has been deleted because topics in Kafka can be unmanaged.

A small extra benefit of using a finalizer is that any errors during deletion can be reported via the `KafkaTopic`'s `status`.
```yaml
status:
  conditions:
  - type: Ready
    status: False
    reason: KafkaError
    message: Deletion failed: ${Error_Message}
    lastTransitionTime: 20230301T103000Z
```

While the behaviour described above will be the _default_ it will be possible to opt out of this using the UTO's new `STRIMZI_ADD_FINALIZER` config parameter. 

* `STRIMZI_USE_FINALIZER=true`: The presence the `strimzi.io/topic-operator` finalizer will be checked (and added, if missing) on every reconciliation where the `metadata.deletionTimestamp` isn't set, including unmanaged topics. The operator will not add or remove the finalizer on `KafkaTopics` which do not match its label selector.
* `STRIMZI_USE_FINALIZER=false`: The operator will remove the `strimzi.io/topic-operator` finalizer if present. 

---

> #### Example 3: `KafkaTopic` deletion propagates to Kafka

Reviewer: Example 2 seems to be missing (=> the numbering is wrong)

> 
> 
> 3. The user can then delete the resource.
>     Because of the presence of a `metadata.finalizer` the resource is not actually deleted.
>     Instead Kube adds the `metadata.deletionTimestamp` field.
> 
>     ```yaml=
>     metadata:
>       generation: 124
>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>       annotations:
>       - strimzi.io/managed: false # do not synchronize to Kafka
>       deletionTimestamp: 20230301T000000.000

>     spec:
>       # ...
>     status:
>       observedGeneration: 124 # == generation => it is not syncrhonized
>     ```
> 
> 4. The operator notices the `metadata.deletionTimestamp` field.
>     Since the `spec.magnaged` is `False` it does **not** attempt topic deletion.
>     1. It removes `strimzi.io/topic-operator` from the `metadata.finalizer`.
>     2. Kube proceeds to remove the resource.

Note that `strimzi.io/managed: false` also means that a `KafkaTopic` can exist without there being a corresponding topic in Kafka.

---

### Operational considerations

#### Changing `metadata.name`

This WILL be supported by:

1. unmanaging the topic (`strimzi.io/managed: false`).
2. deleting the `KafkaTopic`.

> 3. recreating the `KafkaTopic` with a new name.

Reviewer: I guess this assumes there is `.spec.topicName` set. In other words, you can _rename_ this way the KafkaTopic resource, but not the Kafka topic. I think that might be worth clarifying.

#### Change the `metadata.namespace` 

In the case of a UTO configured for multiple namespaces it will be possible to move the `KafkaTopic` for a topic in Kafka from one namespace to another by:

  1. Unmanaging the `KafkaTopic` in the old namespace (as described above)
  2. Deleting the `KafkaTopic` from the old namespace,
  3. Revising the policy (requiring a UTO restart)
  4. Creating a new `KafkaTopic` in the new namespace.

#### Changing `spec.topicName`

This will NOT be supported.
It will be detected by a mismatch between `spec.topicName` and `status.topicName`.
Changes thus detected will result in an error condition.

#### Decreasing `spec.partitions`

This will NOT be supported, because it's not supported by Kafka.
The user will either need to revert the `spec.partitions`, or recreate the topic in Kafka (which they can do via the `KafkaTopic` or not).

#### Changing `spec.replicas`

This will NOT be supported as part of this proposal.

#### Recreating topics in the Kafka cluster

> This WILL be supported.

Reviewer: What exactly does this mean?

Author: Good question. I don't recall now what I meant by it. I'll remove it..

### Managed conflicting rest

* This is the case where there are multiple `KafkaTopics` for the same topic in Kafka, and this one does not have the unique oldest `metadata.creationTimestamp`.
* There may, or may not, be another `KafkaTopic` for the same topic in Kafka in state "Managed conflicting first".
* Changes in this state, including deletions, are not propagated to Kafka.

## Affected/not affected projects

* The existing bidirectional Topic Operator would be deprecated and removed once ZooKeeper-based clusters are no longer supported.
* A new unidirectional Topic Operator would be provided as a replacement, which users could migrate to before ZooKeeper-based clusters are no longer supported.
* The schema of `KafkaTopic` would change
* The schema of `Kafka` would also change (see below).
* The CRD changes would be reflected in the `api` module.

## Compatibility

Let's consider each of the public APIs of the BTO in turn and describe the compatiblity story.

### Operator Configuration

#### Standalone case

| Env var						| Status                        |
|-------------------------------------------------------|-------------------------------|

> | `STRIMZI_NAMESPACE`					|				|

Reviewer 4: I guess it's Unchanged as well, or?

#### CO-managed case

The Topic Operator is usually deployed using the `Kafka` CR (i.e. the Cluster Operator deploys the Topic Operator).
Here is how the schema for the `Kafka` CR will be changed by this proposal:

```yaml
kind: Kafka
metadata:
  name: my-cluster
spec:
  # ...
  topicOperator:

    # Unused options will be deprecated and ignored
    topicMetadataMaxAttempts: 
    zookeeperSessionTimeoutSeconds:
  
    # Options with same schema, but which may need to be reconfigured
    livenessProbe: # probe values may need to be changed
    readinessProbe: # probe values may need to be changed
    startupProbe: # probe values may need to be changed
    resources: # Resource requirements may differ
    jvmOptions: # JVM options may need to be changed
    logging:  # Logger names will need to be changed

    # Unchanged options
    image: 
    watchedNamespace:
    reconciliationIntervalSeconds:
```

### Changes to the KafkaTopic custom resource API

The CR API is _structurally_ unchanged except:

> * `spec.managed` (described above), optional, is added

Reviewer: This is not valid anymore, or?

Furthermore within the same structure:

* `status.conditions` may have new values for the `reason`.

The semantics of the `KafkaTopic` custom resource API will also change:

1. Unidirectional reconciliation
2. The absence of fields will not imply that they take their default values, only that the value is not specified in this CR.

> The reason for 2 is to leave the door open to a further refinement of the UTO in the future.

> > Consider the case where a Kafka cluster is centrally managed by an infrastruture team. 

Reviewer 5: ```suggestion
> Consider the case where a Kafka cluster is centrally managed by an infrastructure team. 
```

  For UTO ApiExceptions from the Admin client will be reported via `Ready` condition.
  For other errors application logging and metrics are considered sufficient.
* For errors when processing a ZooKeeper znode watch.
  UTO doesn't use znode watches, so this cannot happen.
* On attempts to change a `KafkaTopic`'s `spec.topicName`.
  For UTO this will be reported via the status only.
* On detection of conflicting changes between Kube and Kafka.
  For UTO such conflicts cannot happen.
* On attempts to decrease the number of partitions.
  For UTO this will be reported via the status only.

Thus, the UTO will emit no events.
Any software which consumed BTO events will not be compatible with UTO.

## Migration

### Standalone case

Existing users of the standalone BTO would have to:

1. Review how they're using the BTO and whether they require bidirectionality
   1. If bidirectional support is required by their usage then the UTO cannot be used. The user will have no way of managing topics using `KafkaTopic` resources once ZooKeeper support is removed by Apache Kafka and/or Strimzi.
2. Undeploy the BTO (they can retain their existing `KafkaTopics`)

> 3. Deploy the UTO (reconfiguring their `Kafka` CR, or raw `Deployment` in the standalone case). 

Reviewer: This is standalone case -> so the `Kafka` CR mention is not needed here?

|:------|:-----------------------|:-------------------------------------------------------|
| Alpha | 0.35 - 0.37            | Disabled by default                                    |
| Beta  | 0.38 - 0.39            | Enabled by default                                     |
| GA    | 0.40 and newer         | Enabled by default (without possibility to disable it) |

### Disadvantages

#### Use of finalizers

Using finalizers on `KafkaTopics` means that the operator has to be running (to remove the finalizer) before the resources will actually be deleted.
That can prevent other resources from being removed (for example the `Namespace` containing the `KafkaTopics`).
In these cases it is possible to remove the finalizer manually (e.g. via `kubectl edit`), which will then allow the apiserver to remove any other resources.

#### Compatibility with other tooling

This approach is not easily compatible with other topic management tooling.
This includes things like the `kafka-configs.sh` and other Apache Kafka scripts as well as any other tooling which uses the `Admin` client, including Kafka applications themselves.

> In the particular case of a hypthetical "Admin server" that was providing CLI or UI functionality in a Strimzi-aware way, it _might_ be possible to use wrap `Admin` (decorator pattern), or write it using some abstraction over `Admin` which could also support operations on `KafkaTopics` for those interactions which would otherwise result in conflicting sources of truth.

Reviewer 5: ```suggestion
In the particular case of a hypothetical "Admin server" that was providing CLI or UI functionality in a Strimzi-aware way, it _might_ be possible to use wrap `Admin` (decorator pattern), or write it using some abstraction over `Admin` which could also support operations on `KafkaTopics` for those interactions which would otherwise result in conflicting sources of truth.
```

> Alternatively it might be possible to use a Kafka-aware proxy to redirect Kafka protocol messages which changed topic state to the Kube API, along with a Topic Operator which had a priviliged connection that was not proxied in this way.

Reviewer: You say _it might be possible to use a Kafka-aware proxy_ ... but given no proxy with any capability to intercept Kafka topics and manage them through custom resource exists, you it be better to say _it might be possible to develop a Kafka-aware proxy_?

This would result in Kube being the effective source of truth for topic state.

## Future work

> This propsal aims to set the direction and define the semantics of the UTO.

Reviewer 5: ```suggestion
This proposal aims to set the direction and define the semantics of the UTO.
```

It is not intended, by itself, to define the complete post-BTO picture.
Future work _could_ include:

* Providing tooling to conveniently create `KafkaTopics` for existing topics in Kafka.
* Providing tooling to conveniently remove finalizers from `KafkaTopics`

> ## Rejected alternatives

Reviewer: Should you add `Using polling via the Kafka `Admin` client.` as well here? You mention it at the beginning as considered alternative. This is an obvious alternative and people asked about why not do that in the past. So it might be worth listing it here as well.

  It also means that infrastructure-level topics (e.g. `__consumer_offsets` and `__transaction_state`) comingle with per-application topics. 
* It doesn't *enforce* a single source of truth, allowing for changes done directly in Kafka to conflict with declarative deployments. 

Moreover, the BTO doesn't do a good job with item 2. either:

* It only surfaces a subset of topic state through the Kube API (it doesn't include partition leadership/ISR for example)
This is intentional (it doesn't make sense to overload etcd to present this kind of dynamic information), but it undermines the suitability of the BTO for this use case.

For these reasons we are proposing the UTO.

## Proposal

### Goals

* A unidirectional topic operator which is KRaft compatible
* Support for multiple namespaces

### Non-goals

* Supporting RF change (this could be added later)

### CR API

We would remain schema-compatible with the existing `KafkaTopic` resource at the API level. 
This means that all existing `spec` section could be handled by the new operator and all existing fields of the `status` would still be present for applications that consume the status.

> However, the semantics of a `KafkaTopic` would change incompatibly: The `spec` would no longer be updated when Kafka-side configuration changed.

Author: This is correct, actually. While the _syntax_ changes are backwards compatible, the `spec` no longer behaves in exactly the same way.

In other words, the existence of at least some `KafkaTopic` resources would be decoupled from the existence of a corresponding topic in Kafka.
While users could choose to create a topic in Kafka by creating a `KafkaTopic`, they could also create, modify and eventually delete topics directly without a `KafkaTopic` ever existing.

When a `KafkaTopic` did exist for a given topic in Kafka then any changes made via other tooling would eventually be reverted.

### "Managed" and "unmanaged" topics

We will refer to a topic which exists in a Kafka cluster and has a corresponding `KafkaTopic` as a _managed topic_.
A topic which exists in a Kafka cluster without a matching `KafkaTopic` is an _unmanaged topic_.

Naturally any topic created due to the creation of a `KafkaTopic` is managed from the outset.

---

#### Example 1: `KafkaTopic` creation when topic does NOT exist in Kafka

> 1. The user creates a `KafkaTopic`:
> 
>     ```yaml=
>     metadata:
>       generation: 1
>       name: foo
>     spec:
>       partitions: 12
>       replicas: 50
>    ```
> 

> > 2. The operator checks its in-memory map to see whether `foo` is already managed by some other resource.

Author: You're right that this example isn't completely clear. I'll try to reword it.

But to explain here: The operator first checks to see whether there is another resource trying to manage the same topic name, before it even knows whether a topic of that name exists in Kafka. For this example, we're describing the overall outcome where no topic of that name exists in Kafka, and here considering whether there are two resources  trying to manage the same topic.

> 

> >     1. If so, then both resources are updated with error conditions in their `status`.

Author: I forgot to update this bit when I changed this from "both get put into `ResourceConflict`" (which is what I originally had) to "only the newest resource gets put into `ResourceConflict`" aka "first in wins" (which was suggested by @fvaleri). There are pros and cons either way:

Marking both as conflicted:
* means that the topic is temporarily unmanageable via either resource
* means that 
    * either the operator needs to update another resource than the one that changed, complicating the locking which we use to prevent concurrent reconciliation in the case of multi-threaded reconciliation.
    * or that we don't try to update the status of "the other" resource when reconciling "this one", leaving it to the next timed reconciliation to update the other one. The lagginess of this simpler alternative is not attractive.
* means that the error message on the condition needs to enumerate every conflicting resource, which could actually be many more than 2 in the general case.

Marking only the later-created resources is actually simpler, since updates to other resources are avoided and the conflicted resources need only point to the first-created resource. These are all relatively shallow considerations, but together they persuaded me to make the change.

        reason: NotSupported
        message: Decrease of spec.partitions is not supported by Kafka
        lastTransitionTime: 20230301T103000Z
    ```
    The user will then need to decide how to proceed, but typically they might just revert the `spec.partitions`.
* Changes to `spec.replicas` could be supported via CC, but that is not part of _this_ proposal.
    ```yaml
    status:
      conditions:
      - type: Ready
        status: False
        reason: NotSupported
        message: Changing spec.replicas is not supported by the operator
        lastTransitionTime: 20230301T103000Z
    ```

> Note that both these `NotSupported` reasons can arise both from changes being made to the `KafkaTopic`, or changes made directly in Kafka.
> For example, the UTO cannot tell whether the `spec.partitions` was decreased, or whether it was unchanged and the conflict arises because someone increased the number of partitions directly in Kafka so that it looks like the `KafkaTopic` is requesting a decrease. 
> In any case, when the user takes action to fix the problem the `KafkaTopic` will (eventually) get reconciled and the `status` will revert to `Ready`.

### Topic deletion

> We will use a [Kube finalizer](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/) for deletion.

Reviewer: The finalizers still make me nervous. Have you considered the possibility to allow users to disable their use and take the risk the topics might not delete in Kafka when the `KafkaTopic` is deleted?

Reviewer: I was not really questioning the `.spec.managed` here. I'm wondering if we should have a flag to simply disable the finalizers when the user decides so. I.e. if they would be disabled, they would not be set but the user will accept that there might be situations when deleting the `KafkaTopic` does not trigger the deletion in Kafka.

>     ```
> 
> 2. The user then deletes the resource (e.g. `kubectl delete kafkatopic foo`).
>     Because of the presence of a `metadata.finalizer` the resource remains present.
>     Instead Kube adds the `metadata.deletionTimestamp` field.
> 
>     ```yaml=
>     metadata:
>       generation: 124
>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>       deletionTimestamp: 20230301T000000.000
>     spec:
>       # ...
>     status:
>       observedGeneration: 124 # == generation => it is not syncrhonized
>     ```
> 
> 3. The operator notices the `metadata.deletionTimestamp` field.
>    1. If there is more than one `KafkaTopic` which is trying to manage this topic then deletion from Kafka is not attempted;
>       the operator will remove the finalizer and Kube will remove the resource; 
>       upon timed reconciliation the condition on the other `KafkaTopic` will be removed and reconciliation via that other `KafkaTopic` will proceed as normal.
>    2. If deletion is disallowed via `delete.topic.enable=false` broker configuration then the operator will remove the finalizer and Kube will remove the resource. 

> >       The topic in Kafka will thus become unmanaged.

Author: As above, I don't think we need to specify that here.

>    3. Otherwise, the operator attempts to deletes the topic from Kafka.
>        1. If deletion succeeds the operator will remove `strimzi.io/topic-operator` from the `metadata.finalizer` and Kube will remove the resource.
>        2. If deletion failed with `UNKNOWN_TOPIC_OR_PARTITION` (i.e. the topic didn't exist) the operator will continue to remove the finalizer. I.e. this case is not treated as an error.
>        3. If deletion failed with any other error this is reported via a condition in the `status.conditions` and the finalizer is not removed.
---

Similarly to how it is possible to "manage" an existing unmanaged topic in Kafka, we will also provide a mechanism for "unmanaging" a managed topic, via a `spec.managed` flag.
This will allow a `KafkaTopic` to be deleted **without** the topic in Kafka being deleted.

> This can be useful operationally, for example to change the `metadata.name` of a `KafkaTopic`.

Reviewer: Have you considered removing the finalizer from a topic with `spec.managed: false`?

Author: Yes. If a CR with `spec.managed: false` gets deleted then the finalizer would be removed by the operator.

050-states.png shows all the different paths between the different states of the CR.

Note that, as currently specified, there is path where a CR can get deleted without the finalizer being removed by the operator. It happens if the operator's label selector initially matches the CR (so that the finalizer gets added), but the CR is then changed so that it no longer matches the operator's selector (the operator interprets this as "it's not my resource any more", assuming some other operator instance is now managing the CR. But if no other operator takes ownership then when the CR is then deleted there is nothing to delete the finalizer.

Reviewer: > Yes. If a CR with spec.managed: false gets deleted then the finalizer would be removed by the operator.

What I meant here was if you considered removing the finalizer already when the `spec.managed` is set to `false` and not just when it is deleted.

> Note that, as currently specified, there is path where a CR can get deleted without the finalizer being removed by the operator. It happens if the operator's label selector initially matches the CR (so that the finalizer gets added), but the CR is then changed so that it no longer matches the operator's selector (the operator interprets this as "it's not my resource any more", assuming some other operator instance is now managing the CR. But if no other operator takes ownership then when the CR is then deleted there is nothing to delete the finalizer.

This would actually mean such resource cannot be deleted rather then `can get deleted without the finalizer being removed` as the finalizer might stay there. When the labels change and no longer match the selector, the TO will get a DELETE event. So I wonder if we should unset the finalizer at this event. But I guess that might raise conflicts with other instances and would not guarantee anything anyway. So maybe we have to live with it.

> An alternative to `spec.managed` would be "explicit deletion" via a `spec.delete` flag.
> This alternative was rejected because it's incompatible with gitops.

---

#### Example 4: `KafkaTopic` deletion WITHOUT propagation to Kafka

> 1. Assume the topic is currently Kube-managed
> 
>     ```yaml=
>     metadata:
>       generation: 123
>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>     spec:
>       # ...
>     status:
>       observedGeneration: 123 # == generation => it is Kube-managed
>     ```
> 
> 2. The user changes the new `spec.managed` field to `False` so that the topic is no longer Kube-managed.
> 
>     ```yaml=
>     metadata:
>       generation: 124
>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>     spec:
>       managed: False # do not synchronize to Kafka
>     status:
>       observedGeneration: 123 # != generation => it is Kube-managed
>     ```
> 

> > 3. The user waits until the operator has observed that change, indicated by the `status.observedGeneration` matching the `metadata.generation`:

Reviewer: For the paused reconciliation, we also set a `Paused` condition. I think we should do that here as well. It might simplify some of the automation around this and confirm that it is really unmanaged.

>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>     spec:
>       managed: False # do not synchronize to Kafka
>     status:
>       observedGeneration: 124 # == generation => it is now Kube-managed
>     ```
> 
> 4. The user can then delete the resource.
>     Because of the presence of a `metadata.finalizer` the resource is not actually deleted.
>     Instead Kube adds the `metadata.deletionTimestamp` field.
> 
>     ```yaml=
>     metadata:
>       generation: 124
>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>       deletionTimestamp: 20230301T000000.000
>     spec:
>       managed: False # do not synchronize to Kafka
>     status:
>       observedGeneration: 124 # == generation => it is not syncrhonized
>     ```
> 
> 5. The operator notices the `metadata.deletionTimestamp` field.
>     Since the `spec.magnaged` is `False` it does **not** attempt topic deletion.
>     1. It removes `strimzi.io/topic-operator` from the `metadata.finalizer`.
>     2. Kube proceeds to remove the resource.

Note that `spec.managed=False` means that a `KafkaTopic` can exist without there being a corresponding topic in Kafka.

---

> ### Multi-namespace support

Reviewer: The proposal mostly looks good to me. But I think that this is a bit premature to be included in it. I think the multi-namespace support needs to be considered and designed in the context of the UO as well.

* It would be weird to have multi-namespace TO and not UO => it should be probably added in parallel
* It would be good to consider how the namespacing is done in TO and UO. In the UO, in the past, we considered following Kubernetes and their naming principles by suffixing the user name with the namespace as a suffix. This approach might not be suitable for TO and we might intentionally choose a different approach for each of them. But we should do so intentionally after consideration and not arrive at it randomly.
* The policy seems to introduce a parallel security aspect (parallel to Kafka ACLs / UO and to Kube RBAC). It seems to be very limited (managing topics through KafkaTopic resources) while not impacting in any way the other security parts - especially the Kafka ACLs. So, a `team.a` is seemingly an owner of topics `foo-app.*`. But `team.b` can easily create users to produce or consumer to / from these topics. It can even create users with ACLs to delete or modify them. I think this would need to be properly thought through and linked to the UO and ACL management to create some value. Doing it without it can prevent doing this in the future without compatibility issues.

That all seems to be way above this proposal. I would probably prefer to do one of the following:
* Remove the multi-namespace support from this proposal and leave it for later
* Adapt the multi-namespace support to give us more options for the future. E.g.
    * Remove the policy concept until it is properly thought through and deal with multi-namespace conflicts in the same way as for single-namespace conflicts for the time being.
    * Acknowledge the possible future differences how UO might handle multi-namespace support.

----

Assuming you would want to proceed with this as is, I think it is missing clarification on how the multi-namespace watching will be configured. And what the _default_ policy will be. Does the policy configure that? Does the example policy below mean that only namespaces `team.a`, `team.b` and `kafka.cluster.admins` will be watched? That would be inefficient as the number of watches would not scale well when using per-namespace watch.

Reviewer: Perhaps calling it a tenancy model would be more matching than a security model as I referred to it in the initial comment. But I think the core of the comment stands.

> Therefore a different `reason` seems appropriate.

### Operational considerations

#### Changing `metadata.name`

This WILL be supported by:

1. unmanaging the topic (`spec.managed=false`).
2. deleting the `KafkaTopic`.
3. recreating the `KafkaTopic` with a new name.

#### Change the `metadata.namespace` 

In the case of a UTO configured for multiple namespaces it will be possible to move the `KafkaTopic` for a topic in Kafka from one namespace to another by:

  1. Unmanging the `KafkaTopic` in the old namespace (as described above)
  2. Deleting the `KafkaTopic` from the old namespace,
  3. Revising the policy (requiring a UTO restart)
  4. Creating a new `KafkaTopic` in the new namespace.

#### Changing `spec.topicName`

This will NOT be supported.
It will be detected by a mismatch between `spec.topicName` and `status.topicName`.
Changes thus detected will result in an error condition.

#### Decreasing `spec.partitions`

This will NOT be supported, because it's not supported by Kafka.
The user will either need to revert the `spec.partitions`, or recreate the topic in Kafka (which they can do via the `KafkaTopic` or not).

#### Changing `spec.relicas`

> This will NOT be supported using as part of this proposal.

Author: That would be nice, but _this_ proposal guarantee that.

<tr><td>STRIMZI_TLS_ENABLED				<td>Unchanged
<tr><td>STRIMZI_TRUSTSTORE_LOCATION			<td>Unchanged
<tr><td>STRIMZI_TRUSTSTORE_PASSWORD			<td>Unchanged
<tr><td>STRIMZI_KEYSTORE_LOCATION			<td>Unchanged
<tr><td>STRIMZI_KEYSTORE_PASSWORD			<td>Unchanged
<tr><td>STRIMZI_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM	<td>Unchanged
<tr><td>STRIMZI_SASL_ENABLED				<td>Unchanged
<tr><td>STRIMZI_SASL_MECHANISM				<td>Unchanged
<tr><td>STRIMZI_SASL_USERNAME				<td>Unchanged
<tr><td>STRIMZI_SASL_PASSWORD				<td>Unchanged
<tr><td>STRIMZI_SECURITY_PROTOCOL			<td>Unchanged
<tr><td>STRIMZI_ZOOKEEPER_CONNECT			<td>Dropped
<tr><td>STRIMZI_ZOOKEEPER_SESSION_TIMEOUT_MS		<td>Dropped
<tr><td>TC_ZK_CONNECTION_TIMEOUT_MS			<td>Dropped
<tr><td>STRIMZI_REASSIGN_THROTTLE			<td>Dropped, see note 2
<tr><td>STRIMZI_REASSIGN_VERIFY_INTERVAL_MS		<td>Dropped, see note 2
<tr><td>STRIMZI_TOPIC_METADATA_MAX_ATTEMPTS		<td>Dropped
<tr><td>STRIMZI_TOPICS_PATH				<td>Dropped
<tr><td>STRIMZI_STORE_TOPIC				<td>Dropped
<tr><td>STRIMZI_STORE_NAME				<td>Dropped
<tr><td>STRIMZI_APPLICATION_ID				<td>Dropped
<tr><td>STRIMZI_STALE_RESULT_TIMEOUT_MS			<td>Dropped
<tr><td>STRIMZI_USE_ZOOKEEPER_TOPIC_STORE		<td>Dropped

> </table>

Reviewer: Markdown table might be more readable?

2. Although these are present in the BTO source code they have never actually been used.

It is also anticiapted that during development we may want to add support for additional environment variables to configure non-functional aspects of the UTO (e.g. performance-related options like numbers of threads used for processing, etc).

#### CO-managed case

The Topic Operator is usually deployed using the `Kafka` CR (i.e. the Cluster Operator deploys the Topic Operator).
Here is how the schema for the `Kafka` CR will be changed by this proposal:

```yaml
kind: Kafka
spec:
  # ...
  topicOperator:

    # Unused options will be deprecated and ignored
    topicMetadataMaxAttempts: 
    zookeeperSessionTimeoutSeconds:
  
    # Options with same schema, but which may beed to be reconfigured
    livenessProbe: # probe values may need to be changed
    readinessProbe: # probe values may need to be changed
    startupProbe: # probe values may need to be changed
    resources: # Resource requirements may differ
    jvmOptions: # JVM options may need to be changed
    logging:  # Logger names will need to be changed

    # Unchanged options
    image: 

>     watchedNamespace: # becomes optional

Reviewer: This is already optional -> defaults to the namespace of the Kafka CR.

> This proposal doesn't attempt to answer that, since it's not proposed to implement it currently.
> But in order to allow this possibility without introducing an incompatible change in the future we need to define the semantics of fields that are unspecified in the `spec` to mean "unspecified by this CR" rather than "takes the default value".

### Metrics

The BTO exports the following metrics

<table>
<tr><th>Metric type	<th>Metric name				<th>Status
<tr><td>counter		<td>strimzi.reconciliations.periodical	<th>Unchanged
<tr><td>counter		<td>strimzi.reconciliations		<th>Unchanged
<tr><td>counter		<td>strimzi.reconciliations.failed	<th>Unchanged
<tr><td>counter		<td>strimzi.reconciliations.successful	<th>Unchanged
<tr><td>gauge		<td>strimzi.resources			<th>Unchanged
<tr><td>timer		<td>strimzi.reconciliations.duration	<th>Unchanged
<tr><td>gauge		<td>strimzi.reconciliations.paused	<th>Unchanged
<tr><td>counter		<td>strimzi.reconciliations.locked	<th>Unchanged
<tr><td>gauge		<td>strimzi.resource.state		<th>Unchanged

> </table>

Reviewer: Markdown table?

### Emitted Kube `Events`

Let's consider all the cases in which the BTO emits `Event` resources to the Kube apiserver:

* For errors when accessing the topic store. 
  For UTO there is no topic store, so this cannot happen.
* For errors when creating partitions. 
  For UTO ApiExceptions from the Admin client will be reported via `Ready` condition.
  For other errors application logging and metrics are considered sufficient.
* For errors when changing the configuration of topics in Kafka
  For UTO ApiExceptions from the Admin client will be reported via `Ready` condition.
  For other errors application logging and metrics are considered sufficient.
* For errors when processing a ZooKeeper znode watch.
  UTO doesn't use znode watches, so this cannot happen.
* On attempts to change a `KafkaTopic`'s `spec.topicName`.
  For UTO this will be reported via the status only.
* On detection of conflicting changes between Kube and Kafka
  For UTO such conflicts cannot happen.
* On attempts to decrease the number of partitions.
  For UTO this will be reported via the status only.

Thus, the UTO will emit no events.
Any software which consumed BTO events will not be compatible with UTO.

> ## Migration

Reviewer: I think the standalone BTO to UTO migration is quite clear. But the migration within a Kafka cluster would IMHO deserve more details. Will a feature gate drive it? Will there be a flag? Annotation?

To migrate, existing BTO users would have to:

1. Review how they're using the BTO and whether they require bidirectionality

>    1. If bidirectional support is required by their usage then the UTO cannot be used. 

Reviewer: Please clarify here that we will not support the BTO going forward.

> 2. Undeploy the BTO (they can retain their existing `KafkaTopics`)

Author: It's optional and defaults to true, so it's not necessary to add it.

3. Deploy the UTO (reconfiguring their `Kafka` CR, or raw `Deployment` in the standalone case). 
4. Some reconfiguration of pod resources and JVM options may also be required. 

### Disadvantages

#### Use of finalizers

Using finalizers on `KafkaTopics` means that the operator has to be running (to remove the finalizer) before the resources will actually be deleted.
That can prevent other resources from being removed (for example the `Namespace` containing the `KafkaTopics`).
In these cases it is possible to remove the finalizer manually (e.g. via `kubectl edit`), which will then allow the apiserver to remove any other resources.

#### Compatibility with other tooling

This approach is not easily compatible with other topic management tooling. This includes things like the `kafka-configs.sh` and other Apache Kafka scripts as well as any other tooling which uses the `Admin` client, including applications themselves.

<!--Even using the "scheduled auto creation" option, changes done via the admin server would be reverted by the next timed reconciliation. -->

> In the particular case of an "Admin server", to provide CLI or UI functionality in a Strimzi-aware way it _might_ be possible to use wrap `Admin` (decorator pattern), or write it using some abstraction over `Admin` which could also support operations on `KafkaTopics` for those interactions which would otherwise result in conflicting sources of truth.

Reviewer: There is no _Admin server_ in Strimzi. I think this should be removed or be more generic. The Admin server was removed / archived long time ago with the UI.

Alternatively it might be possible to use a Kafka-aware proxy to redirect Kafka protocol messages which changed topic state to the Kube API, along with a Topic Operator which had a priviliged connection that was not proxied in this way. This would result in Kube being the effective source of truth for topic state.

## Future work

This propsal aims to set the direction and define the semantics of the UTO.
It is not intended, by itself, to define the complete post-BTO picture.
Future work _could_ include:

* Providing tooling to conveniently create `KafkaTopics` for existing topics in Kafka.

> * Adopting the config partitioning idea to allow multiple `KafkaTopics` to manage disjoint subsets of topics' configurations.

Reviewer: Possibly you can add here tooling to remove the finalizers?

>     metadata:
>       generation: 124
>       name: foo
>       finalizer: 
>       - strimzi.io/topic-operator
>       deletionTimestamp: 20230301T000000.000
>     spec:
>       managed: False # do not synchronize to Kafka
>     status:
>       observedGeneration: 124 # == generation => it is not syncrhonized
>     ```
> 
> 5. The operator notices the `metadata.deletionTimestamp` field.
>     Since the `spec.magnaged` is `False` it does **not** attempt topic deletion.
>     1. It removes `strimzi.io/topic-operator` from the `metadata.finalizer`.
>     2. Kube proceeds to remove the resource.

Note that `spec.managed=False` also means that a `KafkaTopic` can exist without there being a corresponding topic in Kafka.

---

<!--

### Multi-namespace support

The UTO will support multiple namespaces, subject to the _single resource principal_ [described above](#the-single-resource-principal).

In order to avoid collisions where two or more `KafkaTopics` are trying to manage the same topic in Kafka the operator will enforce a policy for which namespaces can manage which topics. 

---

#### Example

> Conceptually (i.e. this may or may not be configured via YAML):
> 
> ```yaml=

> > policy:

Author: Note that this multi-namespace support is commented out.

> > policy:

Author: I guess I'll remove it properly so reviewers don't get confused.
